src/mlpack/core/data/split_data.hpp

```diff
@@ -127,24 +187,50 @@ void StratifiedSplit(const arma::Mat<T>& input,
 
   trainData.set_size(input.n_rows, trainSize);
   testData.set_size(input.n_rows, testSize);
-  trainLabel.set_size(trainSize);
-  testLabel.set_size(testSize);
+  trainLabel.set_size(inputLabel.n_rows, trainSize);
+  testLabel.set_size(inputLabel.n_rows, testSize);
 
-  for (arma::uword i : order)
+  if (shuffleData)
   {
-    U label = inputLabel[i];
-    if (testLabelCounts[label] < floor(labelCounts[label] * testRatio))
+    arma::uvec order = arma::shuffle(
+        arma::linspace<arma::uvec>(0, input.n_cols - 1, input.n_cols));
+
+    for (arma::uword i : order)
     {
-      testLabelCounts[label] += 1;
-      testData.col(testIdx) = input.col(i);
-      testLabel[testIdx] = inputLabel[i];
-      testIdx += 1;
+      typename LabelsType::elem_type label = inputLabel[i];
+      if (testLabelCounts[label] < floor(labelCounts[label] * testRatio))
+      {
+        testLabelCounts[label] += 1;
+        testData.col(testIdx) = input.col(i);
+        testLabel[testIdx] = inputLabel[i];
+        testIdx += 1;
+      }
+      else
+      {
+        trainData.col(trainIdx) = input.col(i);
+        trainLabel[trainIdx] = inputLabel[i];
+        trainIdx += 1;
+      }
```

Comment:
Up to you---you could also factor this bit into a utility function since it's used below too.  But I'm not sure if it's worth the effort.

Reply:
To refactor this, we would require to pass `order` as an optional argument. I think once I do this for the SplitHelper, then it will be pretty straightforward to do here too

Reply:
Hey @rcurtin , I gave it a thought and I think that we should not change it. I have 2 main reasons for that :
1. The only place where where that piece is used will always remain inside the `StratifiedSplit` function. So, even if there is a duplication of 6-7 lines, then that will always remain together and hopefully won't cause much problem.
2. The implementation of `StratifiedSplit` is pretty straightforward and creating a helper for it will create another level of abstraction.

Do let me know if you agree with my analysis :)

Reply:
Sounds good to me. :+1: